File: all-submissions/0238-product-of-array-except-self/solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int length = nums.length;
        int[] answer = new int[length];

        answer[0] = 1;
        for (int i= 1; i< length; i++) {
            answer[i] = nums[i-1] * answer[i-1];
        }
        int product = nums[length-1];
        for (int i = length-2; i >= 0 ;i--){
            answer[i] = product * answer[i];
            product = product * nums[i];
        }

        return answer;
    }
}

File: all-submissions/0238-product-of-array-except-self/solution.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left_prefix = [1]* n
        right_prefix = [1] * n
        left_prefix[0] = nums[0]
        right_prefix[n-1] = nums[n-1]
        for i in range(1, n):
            left_prefix[i] = left_prefix[i-1] * nums[i]
        for i in range(n-2, -1, -1):
            right_prefix[i] = right_prefix[i+1] * nums[i]
        res = [1]* n
        res[0] = right_prefix[1]
        res[n-1] = left_prefix[n-2]
        for i in range(1, n-1):
            res[i] = left_prefix[i-1] * right_prefix[i+1]
        return res


